### PR TITLE
Support multiple registers and corresponding ctrl values in `CtrlSpec`

### PR DESCRIPTION
**Pull request by tanujkhattar**
`CtrlSpec` currently assumes that the input can be expressed with one register. This PR generalizes the `CtrlSpec` class to get rid of this restriction.

### CODE REVIEW

**Comment by mpharrigan** [comment]
Doesn't this require modifying the `qualtran.Controlled` meta-bloq

**Comment by mpharrigan** [comment]
Answering my own question: no, apparently I wrote it to support this multi-register case already

File: qualtran/_infra/controlled.py [hunk 1]
```diff
@@ -60,17 +72,21 @@ class CtrlSpec:
             are active.
     """
 
-    qdtype: QDType = attrs.field(default=QBit())
-    cvs: NDArray[int] = attrs.field(
-        default=1,
-        converter=lambda cvs: np.array(cvs)
-        if isinstance(cvs, (int, np.integer))
-        else np.asarray(cvs),
+    qdtype: Tuple[QDType, ...] = attrs.field(
```
Comment (mpharrigan) [inline qualtran/_infra/controlled.py:75]: change attribute name to `qdtypes`?

File: qualtran/_infra/controlled.py [hunk 2]
```diff
@@ -60,17 +72,21 @@ class CtrlSpec:
...
-    @property
-    def shape(self) -> Tuple[int, ...]:
-        return self.cvs.shape
+    @cached_property
+    def num_ctrl_reg(self) -> int:
+        return len(self.qdtype)
+
+    @cached_property
+    def shape(self) -> Tuple[Tuple[int, ...], ...]:
+        return tuple(cv.shape for cv in self.cvs)
```
Comment (mpharrigan) [inline qualtran/_infra/controlled.py:89]: rename `shapes`? maybe a little docstring

File: qualtran/_infra/controlled.py (hunk 1 above)
Comment (mpharrigan) [reply qualtran/_infra/controlled.py:75]: anyways: update docstring

**Review by mpharrigan — COMMENTED** [review]
Needs a documentation update

**Comment by mpharrigan** [comment]
Make sure to update the `Controlled.ipynb` documentation as well

File: qualtran/_infra/controlled.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py]: renamed to `qdtypes` and updated docstring

File: qualtran/_infra/controlled.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/_infra/controlled.py]: done

**Comment by tanujkhattar** [comment]
@mpharrigan I've done the renames and updated the documentation. PTAL.

File: qualtran/Controlled.ipynb [hunk 3]
```diff
@@ -54,9 +58,56 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "from qualtran import CtrlSpec\n",
+    "from qualtran import CtrlSpec, QUInt, QBit\n",
     "\n",
-    "ctrl_spec = CtrlSpec(cvs=[0, 1, 0, 1])\n",
+    "ctrl_spec = CtrlSpec(qdtypes=QUInt(4), cvs=0b0110)\n",
+    "assert ctrl_spec.is_active(6)\n",
```
Comment (mpharrigan) [inline qualtran/Controlled.ipynb:64]: this should match the example above where you explicitly note that `0b0101` is integer `5`

**Review by mpharrigan — APPROVED** [review]
please change the notebook example to match the text above it

**Pull request merged**